core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4

```diff
@@ -475,6 +475,63 @@ primaryExpression
     | EXTRACT '(' identifier FROM valueExpression ')'                                     #extract
     | '(' expression ')'                                                                  #parenthesizedExpression
     | GROUPING '(' (qualifiedName (',' qualifiedName)*)? ')'                              #groupingOperation
+    | JSON_EXISTS '(' jsonApi (jsonExistsErrorBehavior ON ERROR)?')'                      #jsonExists
+    | JSON_VALUE '('
+        jsonApi
+        (RETURNING type)?
+        (emptyBehavior=jsonValueBehavior ON EMPTY)?
+        (errorBehavior=jsonValueBehavior ON ERROR)?
+      ')'                                                                                 #jsonValue
+    | JSON_QUERY '('
+        jsonApi
+        (RETURNING type (FORMAT jsonRepresentation)?)?
+        (jsonQueryWrapperBehavior WRAPPER)?
+        ((KEEP | OMIT) QUOTES (ON SCALAR TEXT_STRING)?)?
+        (emptyBehavior=jsonQueryBehavior ON EMPTY)?
+        (errorBehavior=jsonQueryBehavior ON ERROR)?
+      ')'                                                                                 #jsonQuery
+    ;
+
+jsonApi
```

Comment:
Why jsonApi? Also, the PASSING clause seems more related to the function calls above (i.e., why is PASSING here, but not RETURNING?). All in all, it seems like a weird place to "cut" the parsing rules. I would just inline this into each of the rules above -- not much point in sharing this.

Reply:
I think it is a good abstraction, both logically and from the developers viewpoint.
Logically, it encapsulates all that is needed to evaluate the JSON path, that is the path itself plus all the inputs it takes.
And that logic is shared by the three functions implemented in this PR, as well as by other JSON functions yet to be done.
`jsonApi` is the heart of all JSON functions. The other clauses can be considered as "output modifiers", which are specific to particular functions.